### ISSUE TITLE
구매 예정 사용자의 의도 파악

### ISSUE THREAD

**Issue by ocean-park** [codingeverybody/codingyahac#1100]
# 해결하고자 하는 문제
안녕하세요 오션입니다.
저는 게임 개발자 입니다.

한 사용자가 앞으로 아이템을 구매하게 될 지를 AI 를 통해서 예측하고
그 확률을 높히기 위해 쿠폰을 증정하거나 할인을 지원하려는 AI 를 구축하려고 합니다

이런것 들을 예측할 만한 데이터들은 수집할 수 있는데,
사용자들이 시간에 따라 접속 빈도가 높아지거나 하는 것이 기존 사용자들의 구매로 이어지는지를 파악하려면
어떤 방법을 사용해야 할지 힌트를 얻고 싶습니다.

아직 데이터 수집도 하지 않은 상태인데
예를들면 activation 함수를 어떤걸 쓴는게 맞는지
로스 계산 함수는 예제에서 보여주신 mse 를 쓰면 될지
혹은 다른 접근법이 필요한건지 방향을 조언 받고 싶습니다

답변에 감사드립니다

**Comment by JunDamin** [comment]
게임 업계와 쿠팡과 같은 커머스 업계가 달라서 직접적인 도움은 되지 않겠지만 그 접근 방식은 참고할 수 있을 것 같습니다.
[머신러닝을 이용한 실시간 가격 최적화](https://dbr.donga.com/article/view/1401/article_no/8395/ac/magazine)

일단 질문을 구체적으로 정의하는 것부터 시작해야 하지 않을까 싶습니다.
할인과 쿠폰의 효과 보다는 개별적인 개인에게 어떻게 접근해야 할지 고민하시는 듯 합니다.
또한, 구매 여부를 막연하게 두기 보다는 일정 기간을 기준으로 해서 해당 기간 내에 어떤 재화를 구입하는가 여부를 추측하는 것이 아닐까 싶습니다.

그렇게 본다면 일단 데이터를 구분할 수 있을 것 같은데요.
 ▲Input은 사용자 정보와 함께 각 아이템별 쿠폰/할인 일 것이고 ▲Output은 아이템 별 구매 확률이 될 것 같습니다.
여기서 어려운 점은 어떤 사용자 정보를 인풋으로 잡는가 하는 점입니다. 
활발하게 다른 사용자와 상호작용을 하는지, 얼마나 자주 들어오는지, 과거에 아이템을 쓴 적이 있는지 등 고려 할 수 있는 종류는 한도 끝도 없을 것 입니다. 그래서 어떤 데이터를 모아야 하는지 정리하는 것이 가장 먼저 고민하셔야 할 부분일 듯 합니다.

구매 패턴과 상관관계가 있는 사용자 정보들은 모두 필요하지 않을까 싶습니다.

Input과 Output이 확정 되었으면 그 이후에는 이번에 배운 Tensorflow로 레이어를 쌓고 여러가지를 실험해 보면 좋지 않을까 싶습니다. 로스 함수로 mse를 쓰시려면 아웃풋이 연속된 변수여야 합니다. 구매 여부(on, off 방식)의 경우에는 mse를 적용하는 것이 적합하지는 않을 것으로 보입니다.
Tensorflow102 강의에서 이런 경우에는 softmax를 사용했던 걸로 기억합니다. 

제가 아는 건 이정도 이고 이 속에서도 틀린 것들이 있을 가능성은 높습니다.
다만 참고가 되셨길 바라며 마치겠습니다.

**Comment by RayleighKim** [comment]
구매를 reward로 둔 강화학습문제로도 보입니다.
음, 아이템 종류가 한 종류 일까요? 
과거 데이터(추천해본적이 있다면) 를 활용한다면 offline강화로 시작을 해보시는것도 가능할 것 같습니다.

**Comment by JeongsangBaek** [comment]
먼저 각 유저 행동관련 로그를 '모두 다' 확보한 뒤에 데이터 분석을 해보시는 것은 어떨까요?

데이터 분석을 통해 앱 첫 실행부터 7일 15일 30일이 지난 유저들 데이터를 일자별 분석하여,
1. 아이템을 처음 구매하게 되는 시점과 해당 박스플롯. 그리고 평균 구매시점, 데일리 세션 빈도, 구매한 아이템 종류 등 확인
2. 첫 구매 시점에서 아이템을 구매한 유저와 구매하지 않은 유저들의 행동 시퀀스 비교분석, behavior flow분석, funnel 분석
3. 아이템을 구매한 유저와 그렇지 않은 유저간의 이탈시점 박스플롯 비교분석
4. 상관분석을 통해 1-3을 통해 얻은 데이터 필드 중 아이템 구매 여부와 상관관계가 높은 피쳐 선택 및 추출

데이터 분석을 통해 피쳐를 확인했다면 해당 피쳐를 가지고 유저가 아이템을 사는 데 가장 중요한 역할(컨버전)을 하는 유저 행동 혹은 그 행동을 방해하는 허들을 찾아보시면 좋겠습니다. 그리고 컨버전에 긍정적인 영향을 줄 수 있는 새로운 가설을 세워보시고, 가설이 참인지 A/B 테스트등을 통해 확인해보시면 좋겠습니다.

만약 테스트를 통해 긍정적 영향을 준 유저들과 그렇지 않은 유저들 간의 컨버전률 차이를 비교해보시고 통계적 p-value가 0.05 미만이 되었다면, 그때부터는 수집한 데이터에서 추출한 피쳐를 기반으로 컨버전을 일으켰던 행동들을 정리하여 행동들의 분포와 특징을 분석해보시고, 그걸 피쳐로 삼아 피쳐에 맞는 모델을 찾아보시면 좋겠습니다. 머신 러닝 모델링은 그 이후에 시작됩니다.

머신러닝 모델이 어떤 피쳐를 입력받아 무엇을 분류하여 어떤 값을 내보내야 하는지 모르는 상황에서는 어떠한 모델도, 어떠한 모델 설정도 좋은 성능을 보여주기 어렵습니다.

**Comment by ocean-park** [comment]
@JunDamin 
답변 감사합니다. 우선 질문을 조금더 구체화 해 보겠습니다
그 후 사용자 데이터를 가지고 시간으로 쪼개 보고, 해당 데이터를 softmax 로 학습해 보겠습니다.
그루폰 기사도 잘 읽어보겠습니다

**Comment by ocean-park** [comment]
@JeongsangBaek
가능한 행동 데이터를 수집 중입니다.
funnel 도 준비 중입니다. behavior flow 는 funnel 과 다른 내용인지 잘 모르겠네요. 해당 키워드로 검색해 보겠습니다
박스플롯을 그려보는게 도식화에 도움이 많이 되겠네요 (데이터 분석에 대해 더 많이 공부가 필요한 것같습니다 ㅠㅠ)

그리고 피처와 컨버젼, 허들들을 눈으로 좀 확인해야 되는거군요.
이부분은 이번에 학습한 오렌지가 도움이 될 것 같습니다.

이 모든 과정 후에 머신러닝이 시작되는지 몰랐습니다 
너무 좋은 조언 감사드립니다.

**Comment by JeongsangBaek** [comment]
behavior flow는 funnel처럼 하나의 행동으로만 가는 게 아니라, 다음 행동들이 여러개일 경우 보기 좋습니다: [참고 링크](https://www.youtube.com/watch?app=desktop&v=hmLT0RaRe18) 
보통 게임에서는 유저가 구매까지 이루는 유저 행동 시퀀스가 여러가지인 경우가 많아 분석에 유용합니다.

저도 현실 문제를 풀기 전에는 그냥 로그에서 추출한 모든 유저행동 및 피쳐를 다 넣고 큰 모델에 넣으면 머신러닝이 알아서 잘 해줄거야라는 생각을 가지고 있었습니다. 다만 그 모델이 좋은 결과를 냈을 때 "왜" 이 결과가 좋게 나왔는가? 과연 이 모델을 신뢰할 수 있는것인가? 라는 질문에 해답을 찾을 수 없었고 아니나 다를까 실제 예측 결과는 형편없었습니다(그런데 70프로 이상이었다는게 딥러닝의 장점인듯도..) 결국 비즈니스에 사용할 수 없었던 기억이 있습니다. 그래서 확률통계 수업을 온라인 코스로 다시 들었습니다.

일반적인 통계 기반 모델로도 충분히 원하시는 일을 할 수 있습니다 다만 데이터 분석을 통해 '하루에 3번 이상 접속하는 구매 예정자가 D7일 이내에 3만원 이상을 다른 게임에서 구매했었다면 이 게임에서도 컨버전이 높더라' 라는 인사이트를 얻었고 이게 게임을 운영하는 시점들마다 계속 바뀌면 어떤 통계 모델이 가장 예측을 잘 하는지 찾아서 학습시켜 쓰는데 한계가 있습니다. 이럴때 딥러닝 모델을 사용하면 데이터 기반 학습으로 학습을 진행한 시점마다 컨버전율이 가장 높은 유저 세션 카운트, 리텐션 윈도우, ARPPU등을 모델이 알아서 찾아줍니다. 요새 클라우드 서비스들은 데이터만 넣어주면 알아서 모델도 데이터에서 최적의 예측을 진행하는 모델로 바꿔주죠.

부족한 답변이지만 조금이나마 도움이 되셨길 바랍니다.

**Comment by ocean-park** [comment]
제가 생각하고 있던것 역시 데이터를 몽땅 넣으면 결과가 나오겠지.. 였는데, 역시나 아니군요.
 통계를 공부해야 한다는 부분과 말씀해주신 리텐션 등 유용한 사용자 정보를 잘 수집해야겠다는 생각이 듭니다.
 behavior flow 역시 저희가 가지지 못한 부분인데 꼭 필요한 정보라 생각 됩니다.

갈 길이 머네요
장문의 내용 정말 감사합니다
